readthedocs/storage/rclone.py

```diff
@@ -0,0 +1,185 @@
+    rclone_bin = "rclone"
+    default_options = [
+        # Number of file transfers to run in parallel.
+        "--transfers=8",
+        # Skip based on checksum (if available) & size, not mod-time & size.
+        "--checksum",
+        "--verbose",
+    ]
+    env_vars = {}
+
+    # pylint: disable=no-self-use
+    def _build_target_path(self, path):
+        """Build the final target path for the remote."""
+        return path
+
+    def build_target(self, path):
+        """
+        Build the proper target using the current remote type.
+
+        We start the remote with `:` to create it on the fly,
+        instead of having to create a configuration file.
+        See https://rclone.org/docs/#backend-path-to-dir.
+
+        :param path: Path to the remote target.
+        """
+        path = self._build_target_path(path)
+        return f":{self.remote_type}:{path}"
+
+    def execute(self, action, args, options=None):
+        """
+        Execute an rclone subcommand.
+
+        :param action: Name of the subcommand.
+        :param list args: List of positional arguments passed the to command.
+        :param list options: List of options passed to the command.
+        """
+        options = options or []
+        command = [
+            self.rclone_bin,
+            action,
+            *self.default_options,
+            *options,
+            "--",
+            *args,
+        ]
+        env = os.environ.copy()
+        env.update(self.env_vars)
+        log.info("Executing rclone command.", command=command)
+        log.debug("env", env=env)
+        result = subprocess.run(
+            command,
+            capture_output=True,
+            env=env,
+            # TODO: Fail or let the caller decide what to do?
+            check=True,
+        )
+        log.debug(
+            "Result.",
+            stdout=result.stdout.decode(),
+            stderr=result.stderr.decode(),
+            exit_code=result.returncode,
+        )
+        return result
+
+    def sync(self, source, destination):
+        """
+        Run the `rclone sync` command.
+
+        See https://rclone.org/commands/rclone_sync/.
+
+        :params source: Local path to the source directory.
+        :params destination: Remote path to the destination directory.
+        """
+        return self.execute("sync", args=[source, self.build_target(destination)])
+
+
+class RCloneLocal(BaseRClone):
+
+    """
+    RClone remote implementation for the local file system.
+
+    Used for local testing only.
+
+    See https://rclone.org/local/.
+
+    :param location: Root directory where the files will be stored.
+    """
+
+    remote_type = "local"
+
+    def __init__(self, location):
+        self.location = location
+
+    def _build_target_path(self, path):
+        return safe_join_fs(self.location, path)
+
+
+class RCloneS3Remote(BaseRClone):
+
+    """
+    RClone remote implementation for S3.
+
+    All secrets will be passed as environ variables.
```

Comment:
```suggestion
    All secrets will be passed to ``rclone`` command as environ variables.
```

Otherwise, it could be confused by the secrets being passed as env vars to the Python class.